src/expr/src/relation/canonicalize.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Utility functions to transform parts of a single `MirRelationExpr`
+//! into canonical form.
+
+use crate::MirScalarExpr;
+
+/// Fuse equivalence classes of a join.
+///
+/// This function makes it so that the same expression appears in only one
+/// equivalence class. It also sorts and dedups the equivalence classes. (See
+/// [`sort_dedup_equivalences`].)
+///
+/// This function should be called whenever a transformation has added
+/// equivalence classes to a join.
+///
+/// [`sort_dedup_equivalences`]: expr::canonicalize::sort_dedup_equivalences
+/// ```rust
+/// use expr::MirScalarExpr;
+/// use expr::canonicalize::fuse_equivalences;
+///
+/// let mut equivalences_to_fuse = vec![
+///     vec![MirScalarExpr::Column(3), MirScalarExpr::Column(5)],
+///     vec![MirScalarExpr::Column(0), MirScalarExpr::Column(3)]
+/// ];
+/// let expected = vec![
+///     vec![MirScalarExpr::Column(0),
+///         MirScalarExpr::Column(3),
+///         MirScalarExpr::Column(5)]
+/// ];
+/// fuse_equivalences(&mut equivalences_to_fuse);
+/// assert_eq!(expected, equivalences_to_fuse)
+/// ````
+pub fn fuse_equivalences(equivalences: &mut Vec<Vec<MirScalarExpr>>) {
```

Comment:
Maybe a nit, but perhaps we should call this (or some method here `canonicalize`). This one does fuse equivalence classes, but it also sorts, dedups, and discards trivial classes. Thought came up in the docs for JoinMapper which asks for classes to be "fused" but imo "canonicalize" is probably the right concept?